core/src/main/java/com/linecorp/armeria/server/Server.java

```diff
@@ -455,8 +455,16 @@ private void doStop(CompletableFuture<Void> future,
         }
 
         private void finishDoStop(CompletableFuture<Void> future) {
-            // TODO(trustin): Add shutdownBlockingTaskExecutorOnStop
-            // TODO(trustin): Count the pending blocking tasks and wait until it goes zero.
+            if (config.shutdownBlockingTaskExecutorOnStop()) {
+                while (gracefulShutdownSupport.pendingResponses() != 0) {
```

Comment:
I think we need to check whether the `blockingTaskExecutor` has a pending job from here and then shutdown it. 

Checking `pendingResponses` seems okay for the requests received by the server, but we need to consider another jobs which are not belonging to the request/response. 
For example, assume that there's a service which:
- submits a new background task to the `blockingTaskExecutor` when a request is received 
- sends a response immediately after the task is submitted

In this case, there's no `pendingResponses` but the `blockingTaskExecutor` may be executing some job.
So, I think you need to update this `if` block. Also, would you try to add a test for this new feature? You can refer to the following file for writing a test.
https://github.com/line/armeria/blob/master/core/src/test/java/com/linecorp/armeria/server/ServerTest.java

Does it make sense? Please let us know if you have a further question. 😃

Reply:
You are all right, but I can not find the method of the interface for that task.

One idea is to graceful shutdown using `shutdown`, `shutdownNow` and `awaitTermination` of `ExecutorService`,  but it is constrained by this code:
https://github.com/line/armeria/blob/738b4f5c683b07fd620d7a8a5026aeb6a410fb36/core/src/main/java/com/linecorp/armeria/server/ServerConfig.java#L156

`blockingTaskExecutor` are wrapped with `InterminableExecutorService` or `ExecutorBasedExecutorService` at `ServerConfig`, and wrapping class implement this code:
```
@Override
public void shutdown() {
    throw new UnsupportedOperationException();
}

@Override
public List<Runnable> shutdownNow() {
    throw new UnsupportedOperationException();
}
```
How can i check pending jobs?

Adding new features is difficult for me, because i am not familiar with Armenia yet. but I can keep trying wtih thanks to your advice. 
Thank you :D

Reply:
That's a good question. How about this:

- Add a package-local method that returns the original executor to `InterminableExecutorService`.
  - If `instanceof InterminableExecutorService`, unwrap it to use the original `ExecutorService`.
- As a safety measure, catch the `UnsupportedOperationException` raised by `shutdown()` and `awaitTermination()` and log the exception.

Makes sense?